File: php/DataStructureBenchmark/structures/BTree.php
<?php

declare(strict_types=1);

class BTreeNode
{
    public bool $leaf;
    public array $keys = [];      // 각 원소: ['key'=>..., 'keyString'=>..., 'value'=>...]
    public array $children = [];

    public function __construct(bool $leaf = true)
    {
        $this->leaf = $leaf;
    }
}

class BTree
{
    private BTreeNode $root;
    private int $t; // minimum degree

    public function __construct(int $t = 3)
    {
        $this->t = $t;
        $this->root = new BTreeNode(true);
    }

    public function insert(string $key, string $keyString, string $value): void
    {
        $record = [
            'key' => $key,
            'keyString' => $keyString,
            'value' => $value,
        ];

        $root = $this->root;

        if (count($root->keys) === 2 * $this->t - 1) {
            $newRoot = new BTreeNode(false);
            $newRoot->children[] = $root;
            $this->splitChild($newRoot, 0);
            $this->root = $newRoot;
            $this->insertNonFull($newRoot, $record);
        } else {
            $this->insertNonFull($root, $record);
        }
    }

    public function search(string $key): ?array
    {
        return $this->searchNode($this->root, $key);
    }

    public function sortByKeyString(): array
    {
        $result = [];
        $this->traverse($this->root, $result);

        usort($result, static function (array $a, array $b): int {
            return strcmp($a['keyString'], $b['keyString']);
        });

        return $result;
    }

    private function searchNode(BTreeNode $node, string $key): ?array
    {
        $current = $node;
        while (true) {
            $i = 0;
            $cnt = count($current->keys);
            while ($i < $cnt && strcmp($key, $current->keys[$i]['key']) > 0) {
                $i++;
            }
            if ($i < $cnt && $current->keys[$i]['key'] === $key) {
                return $current->keys[$i];
            }
            if ($current->leaf) {
                return null;
            }
            $current = $current->children[$i];
        }
    }

    private function splitChild(BTreeNode $parent, int $index): void
    {
        $t = $this->t;
        $fullChild = $parent->children[$index];
        $newNode = new BTreeNode($fullChild->leaf);

        $middleKey = $fullChild->keys[$t - 1];

        $newNode->keys = array_slice($fullChild->keys, $t);
        $fullChild->keys = array_slice($fullChild->keys, 0, $t - 1);

        if (!$fullChild->leaf) {
            $newNode->children = array_slice($fullChild->children, $t);
            $fullChild->children = array_slice($fullChild->children, 0, $t);
        }

        array_splice($parent->children, $index + 1, 0, [$newNode]);
        array_splice($parent->keys, $index, 0, [$middleKey]);
    }

    private function insertNonFull(BTreeNode $node, array $record): void
    {
        $current = $node;
        while (true) {
            $i = count($current->keys) - 1;

            if ($current->leaf) {
                $current->keys[] = $record;
                while ($i >= 0 && strcmp($record['key'], $current->keys[$i]['key']) < 0) {
                    $current->keys[$i + 1] = $current->keys[$i];
                    $i--;
                }
                $current->keys[$i + 1] = $record;
                return;
            }

            while ($i >= 0 && strcmp($record['key'], $current->keys[$i]['key']) < 0) {
                $i--;
            }
            $i++;

            if (count($current->children[$i]->keys) === 2 * $this->t - 1) {
                $this->splitChild($current, $i);
                if (strcmp($record['key'], $current->keys[$i]['key']) > 0) {
                    $i++;
                }
            }

            $current = $current->children[$i];
        }
    }

    private function traverse(BTreeNode $node, array &$result): void
    {
        // BFS로 전체 키 수집 (sortByKeyString에서 어차피 usort하므로 순서 무관)
        $queue = [$node];
        while (!empty($queue)) {
            $current = array_shift($queue);
            foreach ($current->keys as $key) {
                $result[] = $key;
            }
            if (!$current->leaf) {
                foreach ($current->children as $child) {
                    $queue[] = $child;
                }
            }
        }
    }
}

File: java/DataStructureBenchmark/src/structures/BTree/BTree.java
package structures.BTree;

import java.util.*;

public class BTree {
    private BTreeNode root;
    private int t;

    public BTree(int t) {
        this.t = t;
        this.root = new BTreeNode(true);
    }

    public void insert(String key, String keyString, String value) {
        BTreeEntry entry = new BTreeEntry(key, keyString, value);
        BTreeNode r = root;
        if (r.keys.size() == 2 * t - 1) {
            BTreeNode s = new BTreeNode(false);
            s.children.add(r);
            splitChild(s, 0, r);
            root = s;
            insertNonFull(s, entry);
        } else {
            insertNonFull(r, entry);
        }
    }

    private void insertNonFull(BTreeNode node, BTreeEntry entry) {
        int i = node.keys.size() - 1;
        if (node.leaf) {
            node.keys.add(null); // 공간 확보
            while (i >= 0 && entry.key.compareTo(node.keys.get(i).key) < 0) {
                node.keys.set(i + 1, node.keys.get(i));
                i--;
            }
            node.keys.set(i + 1, entry);
        } else {
            while (i >= 0 && entry.key.compareTo(node.keys.get(i).key) < 0) {
                i--;
            }
            i++;
            if (node.children.get(i).keys.size() == 2 * t - 1) {
                splitChild(node, i, node.children.get(i));
                if (entry.key.compareTo(node.keys.get(i).key) > 0) {
                    i++;
                }
            }
            insertNonFull(node.children.get(i), entry);
        }
    }

    private void splitChild(BTreeNode parent, int i, BTreeNode y) {
        BTreeNode z = new BTreeNode(y.leaf);

        // Split keys
        z.keys.addAll(y.keys.subList(t, y.keys.size()));
        y.keys.subList(t, y.keys.size()).clear();

        // Split children if not a leaf
        if (!y.leaf) {
            z.children.addAll(y.children.subList(t, y.children.size()));
            y.children.subList(t, y.children.size()).clear();
        }

        // Add new child and key to parent
        parent.children.add(i + 1, z);
        parent.keys.add(i, y.keys.remove(t - 1));
    }

    public BTreeEntry search(String key) {
        return searchNode(root, key);
    }

    private BTreeEntry searchNode(BTreeNode node, String key) {
        int i = 0;
        while (i < node.keys.size() && key.compareTo(node.keys.get(i).key) > 0) {
            i++;
        }
        if (i < node.keys.size() && node.keys.get(i).key.equals(key)) {
            return node.keys.get(i);
        }
        if (node.leaf) {
            return null;
        }
        return searchNode(node.children.get(i), key);
    }
}